packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -196,54 +227,125 @@ abstract class ImageStreamCompleter extends Diagnosticable {
   /// occurred. If the listener is added within a render object paint function,
   /// then use this flag to avoid calling [RenderObject.markNeedsPaint] during
   /// a paint.
-  void addListener(ImageListener listener) {
-    _listeners.add(listener);
-    if (_current != null) {
+  void addListener(ImageListener listener, { ImageErrorListener onError }) {
+    _listeners.add(new _ImageListenerPair(listener, onError));
+    if (_currentImage != null) {
       try {
-        listener(_current, true);
+        listener(_currentImage, true);
       } catch (exception, stack) {
-        _handleImageError('by a synchronously-called image listener', exception, stack);
+        reportError(
+          context: 'by a synchronously-called image listener',
+          exception: exception,
+          stack: stack,
+        );
       }
     }
+    if (_currentError != null && onError != null) {
+      try {
+          onError(_currentError.exception, _currentError.stack);
+        } catch (exception, stack) {
+          FlutterError.reportError(
+            new FlutterErrorDetails(
+              exception: exception,
+              library: 'image resource service',
+              context: 'by a synchronously-called image error listener',
+              stack: stack,
+            ),
+          );
+        }
+    }
   }
 
-  /// Stop listening for new concrete [ImageInfo] objects.
+  /// Stop listening for new concrete [ImageInfo] objects and errors from
+  /// its associated [ImageErrorListener].
   void removeListener(ImageListener listener) {
-    _listeners.remove(listener);
+    _listeners.removeWhere((_ImageListenerPair listenerPair) {
+      return listenerPair.listener == listener;
+    });
```

Comment:
this removes all the matches, we should only remove the first match.

Reply:
Ah, other `removeListener`s seem to do that. Done in follow-up PR